packages/tizen_app_manager/tizen/src/application_utils.cc

```diff
@@ -4,145 +4,194 @@
+}
+
+void GetForEachMetadata(app_info_h app_info, flutter::EncodableMap &value) {
+  flutter::EncodableMap metadata;
 
-  ret = app_info_foreach_metadata(app_info, AppMetaDataCB, (void *)&metadata);
+  int ret = app_info_foreach_metadata(
+      app_info,
+      [](const char *key, const char *value, void *user_data) {
+        if (key != nullptr) {
+          flutter::EncodableMap *metadata =
+              static_cast<flutter::EncodableMap *>(user_data);
+          metadata->insert(
+              std::pair<flutter::EncodableValue, flutter::EncodableValue>(
+                  std::string(key), std::string(value)));
+          LOG_INFO("AppMetaDataCB key: %s, value: %s", key, value);
+        }
+        return true;
+      },
+      (void *)&metadata);
   if (ret != APP_MANAGER_ERROR_NONE) {
     LOG_ERROR("app_info_foreach_metadata error! : %s", get_error_message(ret));
-    goto cleanup;
-  }
-
-  value.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
-      "appId", std::string(app_id)));
-  value.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
-      "packageId", std::string(pkg_id)));
-  value.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
-      "label", std::string(label)));
-  value.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
-      "type", std::string(type)));
-  if (icon_path != nullptr) {
-    value.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
-        "iconPath", std::string(icon_path)));
+    throw ret;
   }
-  value.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
-      "executablePath", std::string(exec_path)));
-  value.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
-      "sharedResourcePath", std::string(shared_res_path)));
-  value.insert(std::pair<flutter::EncodableValue, flutter::EncodableValue>(
-      "isNoDisplay", no_display));
   value.insert(std::pair<flutter::EncodableValue, flutter::EncodableMap>(
       "metadata", metadata));
+}
+}  // namespace
 
-cleanup:
-  if (app_id) {
-    free(app_id);
-  }
-  if (pkg_id) {
-    free(pkg_id);
-  }
-  if (label) {
-    free(label);
-  }
-  if (type) {
-    free(type);
-  }
-  if (icon_path) {
-    free(icon_path);
+namespace application_utils {
+bool ExtractValueFromMap(const flutter::EncodableValue &arguments,
+                         const char *key, std::string &out_value) {
+  if (std::holds_alternative<flutter::EncodableMap>(arguments)) {
+    flutter::EncodableMap map = std::get<flutter::EncodableMap>(arguments);
+    auto iter = map.find(flutter::EncodableValue(key));
+    if (iter != map.end() && !iter->second.IsNull()) {
+      if (auto pval = std::get_if<std::string>(&iter->second)) {
+        out_value = *pval;
+        return true;
+      }
+    }
   }
-  if (exec_path) {
-    free(exec_path);
+  return false;
+}
+
+int GetAppData(app_info_h app_info, flutter::EncodableMap &value) {
```

Comment:
https://github.com/flutter-tizen/plugins/issues/354

Ideally the type `flutter::EncodableMap` should not appear in this file. Consider moving `ExtractValueFromMap` and `GetAppData` to `tizen_app_manager_plugin.cc`.

Reply:
> The classes should have a clean design so that their roles can be easily known from their signatures. Ideally, the plugin class should be the only one that interacts with the underlying platform channel (processes method arguments and generates results), and other classes should not touch any Flutter API (flutter/*.h).

As you said, GetAppData() and ExtractValueFromMap() were moved to tizen_app_manager_plugin.cc.